sympy/polys/domains/field.py

```diff
@@ -68,6 +68,39 @@ def gcd(self, a, b):
 
         return self.convert(p, ring)/q
 
+    def gcdex(self, a, b):
+        """
+        Returns x, y, g such that a * x + b * y == g == gcd(a, b)
```

Comment:
I'm not sure about defining this as a generic method on `Field` especially if the implementation is different depending on `get_ring`.

We could perhaps have two implementations here like:
```python
def _gcdex_fraction_field(...):
    ring = self.get_ring()
   ...

def _gcdex_generic_field(...):
   ...
```
I think that the various subclasses need to opt-in explicitly to call those methods though and each needs to be considered on a case by case basis.

One case I am not sure about is algebraic fields. SymPy does not implement the ring of integers of an algebraic field:
```python
In [1]: K = QQ.algebraic_field(sqrt(2))

In [2]: K
Out[2]: QQ<sqrt(2)>

In [3]: K.get_ring()
---------------------------------------------------------------------------
DomainError                               Traceback (most recent call last)
Cell In[3], line 1
----> 1 K.get_ring()
```
Really though this is basically analogous to `QQ_I` and so we would to handle it the same sort of way:
```python
In [4]: QQ_I.get_ring()
Out[4]: ZZ_I

In [10]: from sympy.matrices.normalforms import smith_normal_decomp

In [11]: A = randMatrix(3, max=10)/7

In [12]: B = randMatrix(3, max=10)/7

In [13]: smith_normal_decomp(A+I*B, domain=QQ_I)
Out[13]: 
⎛           ⎡    21   7⋅ⅈ                                             ⎤                            ↪
⎜           ⎢    ── - ───               0                   0         ⎥  ⎡     16   3⋅ⅈ    1858    ↪
⎜           ⎢    10   10                                              ⎥  ⎢1  - ── - ───  - ──── -  ↪
⎜⎡1  0  0⎤  ⎢                                                         ⎥  ⎢     5     5     925     ↪
⎜⎢       ⎥  ⎢   119   63⋅ⅈ          28    91⋅ⅈ                        ⎥  ⎢                         ↪
⎜⎢0  1  0⎥, ⎢   ─── - ────        - ─── + ────              0         ⎥, ⎢                  41     ↪
⎜⎢       ⎥  ⎢   185   185           185   185                         ⎥  ⎢0      1        - ─── -  ↪
⎜⎣0  0  1⎦  ⎢                                                         ⎥  ⎢                  185    ↪
⎜           ⎢48181    183715⋅ⅈ  106953   236957⋅ⅈ    75432    103201⋅ⅈ⎥  ⎢                         ↪
⎜           ⎢────── - ────────  ────── - ────────  - ────── + ────────⎥  ⎣0      0             1   ↪
⎝           ⎣721034    721034   360517    360517     360517    360517 ⎦                            ↪

↪       ⎞
↪ 344⋅ⅈ⎤⎟
↪ ─────⎥⎟
↪  925 ⎥⎟
↪      ⎥⎟
↪ 98⋅ⅈ ⎥⎟
↪ ──── ⎥⎟
↪ 185  ⎥⎟
↪      ⎥⎟
↪      ⎦⎟
↪       ⎠

In [15]: smith_normal_decomp(A+B*sqrt(2), domain=QQ.algebraic_field(sqrt(2)))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[15], line 1
----> 1 smith_normal_decomp(A+B*sqrt(2), domain=QQ.algebraic_field(sqrt(2)))

File ~/current/active/sympy/sympy/matrices/normalforms.py:71, in smith_normal_decomp(m, domain)
     57 '''
     58 Return the Smith Normal Decomposition of a matrix `m` over the ring
     59 `domain`. This will only work if the ring is a principal ideal domain.
   (...)
     68 >>> assert a == s * m * t
     69 '''
     70 dM = _to_domain(m, domain)
---> 71 a, s, t = _snd(dM)
     72 return a.to_Matrix(), s.to_Matrix(), t.to_Matrix()

File ~/current/active/sympy/sympy/polys/matrices/normalforms.py:116, in smith_normal_decomp(m)
    113 rows, cols = shape = m.shape
    114 m = m.to_list()
--> 116 invs, s, t = _smith_normal_decomp(m, domain, shape=shape, full=True)
    117 smf = DomainMatrix.diag(invs, domain, shape).to_dense()
    119 s = DomainMatrix(s, domain=domain, shape=(rows, rows))

File ~/current/active/sympy/sympy/polys/matrices/normalforms.py:227, in _smith_normal_decomp(m, domain, shape, full)
    225 c = domain.canonical_unit(m[0][0])
    226 if domain.is_Field:
--> 227     c = 1 / m[0][0]
    228 if c != domain.one:
    229     m[0][0] *= c

TypeError: unsupported operand type(s) for /: 'int' and 'ANP'
```
Probably `ANP` should define `__rdiv__` for int but at the same time it should be `domain.one / m[0][0]` or:
```python
In [16]: K = QQ.algebraic_field(sqrt(2))

In [17]: a = K.from_sympy(1+sqrt(2))

In [18]: 1/a
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[18], line 1
----> 1 1/a

TypeError: unsupported operand type(s) for /: 'int' and 'ANP'

In [19]: a ** -1
Out[19]: ANP([1, -1], [1, 0, -2], QQ)

In [20]: K.one / a
Out[20]: ANP([1, -1], [1, 0, -2], QQ)
```